exchange/auction.go

```diff
@@ -60,111 +60,118 @@ func (a *auction) setRoundedPrices(priceGranularity openrtb_ext.PriceGranularity
 	a.roundedPrices = roundedPrices
 }
 
-func (a *auction) doCache(ctx context.Context, cache prebid_cache_client.Client, bids bool, vast bool, bidRequest *openrtb.BidRequest, ttlBuffer int64, defaultTTLs *config.DefaultTTLs, bidCategory map[string]string) []error {
-	if !bids && !vast {
-		return nil
-	}
-
+func (a *auction) doCache(ctx context.Context, cache prebid_cache_client.Client, targData *targetData, bidRequest *openrtb.BidRequest, ttlBuffer int64, defaultTTLs *config.DefaultTTLs, bidCategory map[string]string) []error {
+	var bids, vast, includeBidderKeys, includeWinners bool = targData.includeCacheBids, targData.includeCacheVast, targData.includeBidderKeys, targData.includeWinners
 	var errs []error
-	expectNumBids := valOrZero(bids, len(a.roundedPrices))
-	expectNumVast := valOrZero(vast, len(a.roundedPrices))
-	bidIndices := make(map[int]*openrtb.Bid, expectNumBids)
-	vastIndices := make(map[int]*openrtb.Bid, expectNumVast)
-	toCache := make([]prebid_cache_client.Cacheable, 0, expectNumBids+expectNumVast)
-	expByImp := make(map[string]int64)
-	competitiveExclusion := false
-	var hbCacheID string
-	if len(bidCategory) > 0 {
-		// assert:  category of winning bids never duplicated
-		if rawUuid, err := uuid.NewV4(); err == nil {
-			hbCacheID = rawUuid.String()
-			competitiveExclusion = true
-		} else {
-			errs = append(errs, errors.New("failed to create custom cache key"))
-		}
-	}
 
-	// Grab the imp TTLs
-	for _, imp := range bidRequest.Imp {
-		expByImp[imp.ID] = imp.Exp
-	}
-	for _, topBidsPerImp := range a.winningBidsByBidder {
-		for _, topBidPerBidder := range topBidsPerImp {
-			impID := topBidPerBidder.bid.ImpID
-			var customCacheKey string
-			var catDur string
-			useCustomCacheKey := false
-			if competitiveExclusion && topBidPerBidder == a.winningBids[impID] {
-				// set custom cache key for winning bid when competitive exclusion applies
-				catDur = bidCategory[topBidPerBidder.bid.ID]
-				if len(catDur) > 0 {
-					customCacheKey = fmt.Sprintf("%s_%s", catDur, hbCacheID)
-					useCustomCacheKey = true
-				}
+	if (bids || vast) && (includeBidderKeys || includeWinners) {
```

Comment:
I don't understand the use of includeBidderKeys and includeWinners in this function:
- purpose of this change is to cache only winners but we will cache non-winners if includeBidderKeys is set
- if neither of these is set, we won't cache winners; this is a change from previous behavior, isn't it?

Reply:
current version of the `addKeys` function found in `exchange/utils.go` also uses this logic to create targets. I don't know if this logic that is currently in place in `addKeys` is correct or makes perfect sense, I just replicated it.

Reply:
If we are setting keys for all bidders, then we need to cache all bidders. But it may be best to preserve the initial logic flow and bail out of the function if we have nothing to do rather than putting everything inside an if statement.

Reply:
done